data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java

```diff
@@ -178,47 +216,35 @@ private void handleRetry(final AccumulatingBulkRequest request, final BulkRespon
                         bulkRequestBadErrors.increment();
                     }
                 }
-
-                if (canRetry(e)) {
-                    handleRetry(bulkRequestForRetry, null, backOffUtils, false);
-                    bulkRequestNumberOfRetries.increment();
-                } else {
-                    handleFailures(bulkRequestForRetry, e);
-                    bulkRequestFailedCounter.increment();
-                }
-
+                handleRetriesAndFailures(bulkRequestForRetry, retryCount, backOffUtils, null, e);
                 return;
             }
             if (bulkResponse.errors()) {
-                if (canRetry(bulkResponse)) {
-                    if (firstAttempt) {
-                        for (final BulkResponseItem bulkItemResponse : bulkResponse.items()) {
-                            if (bulkItemResponse.error() == null) {
-                                sentDocumentsOnFirstAttemptCounter.increment();
-                            }
-                        }
-                    }
-                    handleRetry(bulkRequestForRetry, bulkResponse, backOffUtils, false);
-                } else {
-                    handleFailures(bulkRequestForRetry, bulkResponse.items());
-                }
+                handleRetriesAndFailures(bulkRequestForRetry, retryCount, backOffUtils, bulkResponse, null);
             } else {
                 final int numberOfDocs = bulkRequestForRetry.getOperationsCount();
-                if (firstAttempt) {
+                if (retryCount == 1) {
```

Comment:
It might be good to create a variable `boolean firstAttempt = retryCount == 1`. It is a little unclear in this function that retryCount = 1 actually means that there have been no retries and this is the first attempt.